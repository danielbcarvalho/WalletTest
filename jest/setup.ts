import 'react-native-gesture-handler/jestSetup';
import { mockDefaultTheme } from '../src/theme/defaultTheme';

jest.mock('styled-components', () => {
  const original = jest.requireActual('styled-components');

  return {
    ...original,
    useTheme: () => {
      return mockDefaultTheme;
    },
  };
});
