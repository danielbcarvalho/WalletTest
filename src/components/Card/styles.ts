import { css } from 'styled-components';
import styled from 'styled-components/native';

interface Props {
  kind: 'black' | 'green';
}
export const Container = styled.View<Props>`
  border-radius: 16px;
  display: flex;
  width: 100%;
  height: 180px;
  padding: 20px;
  justify-content: space-around;
  align-items: flex-start;
  gap: 10px;
  background-color: ${({ theme, kind }) =>
    theme.components.card[kind].backgroundColor};
`;

export const TextWrapper = styled.View`
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: flex-start;
`;

export const Title = styled.Text<Props>`
  ${({ theme, kind }) => css`
    font-family: ${theme.components.card.title.fontFamily};
    font-size: ${theme.components.card.title.fontSize};
    line-height: ${theme.components.card.title.lineHeight};
    color: ${theme.components.card[kind].color};
  `};
`;

export const Name = styled.Text<Props>`
  ${({ theme, kind }) => css`
    font-family: ${theme.components.card.name.fontFamily};
    font-size: ${theme.components.card.name.fontSize};
    line-height: ${theme.components.card.name.lineHeight};
    color: ${theme.components.card[kind].color};
  `};
`;

export const Text = styled.Text<Props>`
  ${({ theme, kind }) => css`
    font-family: ${theme.components.card.text.fontFamily};
    font-size: ${theme.components.card.text.fontSize};
    line-height: ${theme.components.card.text.lineHeight};
    color: ${theme.components.card[kind].color};
    text-align: center;
  `};
`;
