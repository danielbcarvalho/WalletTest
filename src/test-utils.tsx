// TODO: Verificar essa melhoria nos testes
import React from 'react';
import {
  render,
  RenderAPI,
  RenderOptions,
} from '@testing-library/react-native';
import renderer, { ReactTestRenderer } from 'react-test-renderer';
import { ThemeProvider } from 'styled-components/native';

import { mockDefaultTheme } from './theme/defaultTheme';

interface IAllThemeProviders {
  children: React.ReactElement;
}
const AllTheProviders = ({ children }: IAllThemeProviders) => {
  return <ThemeProvider theme={mockDefaultTheme}>{children}</ThemeProvider>;
};

function customRender(
  ui: React.ReactElement,
  options = {} as RenderOptions,
): RenderAPI {
  return render(ui, { wrapper: AllTheProviders, ...options });
}

export function rendererCreate(
  component: React.ReactElement,
): ReactTestRenderer {
  return renderer.create(<AllTheProviders>{component}</AllTheProviders>);
}

// re-export everything
export * from '@testing-library/react-native';

// override render method
export { customRender as render };
