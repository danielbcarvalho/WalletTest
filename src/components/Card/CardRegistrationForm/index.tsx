import React from 'react';

import { useTranslation } from 'react-i18next';
import { Masks } from 'react-native-mask-input';
import { yupResolver } from '@hookform/resolvers/yup';
import { useForm, Controller } from 'react-hook-form';
import { useFocusEffect } from '@react-navigation/native';

import { Button } from '../../Button';
import InputCustom from '../../InputCustom';

import { BottomInputWrapper } from './styles';
import { cardFormValidationSchema } from './schema';
import { getMaskCVV, getMaskDateMMYY } from '../../../utils/mask-input';

interface Props {
  onCardRegisterForm: (card: FormValues) => void;
}

interface FormValues {
  number: string;
  name: string;
  expiry: string;
  cvv: string;
}

function CardRegistrationForm({ onCardRegisterForm }: Props) {
  const { t } = useTranslation();
  const {
    control,
    handleSubmit,
    formState,
    reset: clearForm,
  } = useForm<FormValues>({
    defaultValues: {
      number: '',
      name: '',
      expiry: '',
      cvv: '',
    },
    resolver: yupResolver(cardFormValidationSchema),
    mode: 'all',
  });

  const onSubmit = (data: FormValues) => {
    const card: FormValues = {
      name: data.name,
      number: data.number,
      cvv: data.cvv,
      expiry: data.expiry,
    };
    onCardRegisterForm(card);
  };

  useFocusEffect(
    React.useCallback(() => {
      const unsubscribe = () => clearForm();

      return () => unsubscribe();
    }, [clearForm]),
  );

  return (
    <>
      <Controller
        control={control}
        render={({ field: { onChange, onBlur, value } }) => {
          return (
            <InputCustom
              accessibilityLabel={t('card number')}
              onBlur={onBlur}
              onChangeText={onChange}
              value={value}
              label={t('card number')}
              placeholder=""
              keyboardType="numeric"
              mask={Masks.CREDIT_CARD}
              name="number"
            />
          );
        }}
        name="number"
      />
      <Controller
        control={control}
        render={({ field: { onChange, onBlur, value } }) => (
          <InputCustom
            accessibilityLabel={t('card holder name')}
            onBlur={onBlur}
            onChangeText={onChange}
            value={value}
            label={t('card holder name')}
            keyboardType="default"
            autoCapitalize="words"
            name="name"
          />
        )}
        name="name"
      />
      <BottomInputWrapper>
        <Controller
          control={control}
          render={({ field: { onChange, onBlur, value } }) => (
            <InputCustom
              accessibilityLabel={t('expiration date')}
              placeholder="00/00"
              onBlur={onBlur}
              onChangeText={onChange}
              value={value}
              label={t('expiration date')}
              keyboardType="numeric"
              width={48}
              mask={getMaskDateMMYY}
              name="expiry"
            />
          )}
          name="expiry"
        />
        <Controller
          control={control}
          render={({ field: { onChange, onBlur, value } }) => (
            <InputCustom
              accessibilityLabel={t('cvv')}
              placeholder="***"
              onBlur={onBlur}
              onChangeText={onChange}
              value={value}
              label={t('cvv')}
              keyboardType="numeric"
              width={48}
              mask={getMaskCVV}
              name="cvv"
            />
          )}
          name="cvv"
        />
      </BottomInputWrapper>
      <Button
        title={t('advance')}
        accessibilityLabel={t('advance')}
        onPress={handleSubmit(onSubmit)}
        disabled={!formState.isValid}
      />
    </>
  );
}

export default CardRegistrationForm;
