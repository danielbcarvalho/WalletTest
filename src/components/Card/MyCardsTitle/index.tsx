import React from 'react';
import { useTranslation } from 'react-i18next';
import { useTheme } from 'styled-components/native';

import { H4 } from '../../Typografy/H4';

import { Container } from './styles';

function MyCardExtendedHeader() {
  const theme = useTheme();
  const { t } = useTranslation();
  return (
    <Container>
      <H4 color={theme.pallete.info.main}>{t('My cards')}</H4>
    </Container>
  );
}

export default MyCardExtendedHeader;
