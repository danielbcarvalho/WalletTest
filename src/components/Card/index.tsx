import React from 'react';
import { useTranslation } from 'react-i18next';

import { Card as CardModel } from '../../models/CardModels';

import { Container, Text, Title, Name, TextWrapper } from './styles';

function Card({
  card,
  kind = 'black',
}: {
  card: CardModel;
  kind?: 'black' | 'green';
}) {
  const { t } = useTranslation();
  return (
    <Container kind={kind}>
      <Title kind={kind}>
        {kind === 'black' ? 'Black Card' : 'Green Card'}
      </Title>
      <TextWrapper>
        <Name kind={kind}>{card.name}</Name>
        <Text kind={kind}>{`**** **** **** ${card.number.slice(-4)}`}</Text>
        <Text kind={kind}>{`${t('validity')} ${card.expiry}`}</Text>
      </TextWrapper>
    </Container>
  );
}

export default Card;
