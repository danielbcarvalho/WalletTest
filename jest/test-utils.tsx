import React, { ReactElement, FC } from 'react';
import { render } from '@testing-library/react-native';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { ThemeProvider } from '../src/theme';

type Options = Parameters<typeof render>[1];

const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      cacheTime: Infinity,
    },
  },
});
const AllTheProviders: FC<{ children: React.ReactNode }> = ({ children }) => (
  <ThemeProvider>
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  </ThemeProvider>
);

const customRender = (ui: ReactElement, options?: Options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react-native';
export { customRender as render };
