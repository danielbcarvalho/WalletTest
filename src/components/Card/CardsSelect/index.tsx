import React from 'react';
import { useTranslation } from 'react-i18next';

import Card from '..';
import { Card as CardModel } from '../../../models/CardModels';

import {
  Text,
  Container,
  UseButton,
  CardFlatList,
  CardContainer,
} from './styles';
import { useCardList } from '../../hooks/useCardList';
import { isCardOnTheTop } from '../../../utils/card';

function CardsSelect() {
  const { t } = useTranslation();

  const { cards, setCardOnTop, onUseThisCard } = useCardList();

  if (cards?.length === 0) {
    return <Text>{t('ops, no card registered')}</Text>;
  }

  function renderUseCardButton() {
    return (
      <UseButton onPress={onUseThisCard}>
        <Text>{t('use this card')}</Text>
      </UseButton>
    );
  }

  function renderCard({ item, index }: { item: CardModel; index: number }) {
    return (
      <CardContainer
        disabled={isCardOnTheTop(index, cards)}
        activeOpacity={0.7}
        cardTop={isCardOnTheTop(index, cards)}
        onPress={() => setCardOnTop(item.id)}>
        <Card card={item} />
      </CardContainer>
    );
  }

  return (
    <Container>
      <CardFlatList
        data={cards}
        bounces={false}
        renderItem={renderCard}
        ListFooterComponent={renderUseCardButton}
        keyExtractor={item => item.id.toString()}
      />
    </Container>
  );
}

export default CardsSelect;
