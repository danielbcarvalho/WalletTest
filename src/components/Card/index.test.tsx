import React from 'react';

import { render } from '../../../jest/test-utils';

import Card from '.';

describe('Card', () => {
  const card = {
    id: '4ec42ba9-50af-40d2-af90-8312edbd9ca2',
    number: '3529 5435 3355 8727',
    cvv: '317',
    name: 'John Doe',
    expiry: '12/24',
  };

  test('renders card with black kind', () => {
    const { getByText } = render(<Card card={card} kind="black" />);
    expect(getByText('Black Card')).toBeTruthy();
    expect(getByText('John Doe')).toBeTruthy();
    expect(getByText('**** **** **** 8727')).toBeTruthy();
    expect(getByText('validity 12/24')).toBeTruthy();
  });

  test('renders card with green kind', () => {
    const { getByText } = render(<Card card={card} kind="green" />);
    expect(getByText('Green Card')).toBeTruthy();
    expect(getByText('John Doe')).toBeTruthy();
    expect(getByText('**** **** **** 8727')).toBeTruthy();
    expect(getByText('validity 12/24')).toBeTruthy();
  });
  test('matches snapshot', () => {
    const cardMock = {
      id: '4ec42ba9-50af-40d2-af90-8312edbd9ca2',
      number: '3529 5435 3355 8727',
      cvv: '317',
      name: 'John Doe',
      expiry: '12/23',
    };
    const { toJSON } = render(<Card card={cardMock} />);
    expect(toJSON()).toMatchSnapshot();
  });
});
