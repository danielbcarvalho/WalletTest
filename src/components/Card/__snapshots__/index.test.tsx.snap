// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card matches snapshot 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "flex-start",
      "backgroundColor": "#000000",
      "borderBottomLeftRadius": 16,
      "borderBottomRightRadius": 16,
      "borderTopLeftRadius": 16,
      "borderTopRightRadius": 16,
      "display": "flex",
      "gap": 10,
      "height": 180,
      "justifyContent": "space-around",
      "opacity": 1,
      "paddingBottom": 20,
      "paddingLeft": 20,
      "paddingRight": 20,
      "paddingTop": 20,
      "width": "100%",
    }
  }
>
  <Text
    kind="black"
    style={
      {
        "color": "#FFFFFF",
        "fontFamily": "PTSans-Regular",
        "fontSize": 18,
        "lineHeight": 20,
      }
    }
  >
    Black Card
  </Text>
  <View
    style={
      {
        "alignItems": "flex-start",
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "flex-start",
      }
    }
  >
    <Text
      kind="black"
      style={
        {
          "color": "#FFFFFF",
          "fontFamily": "PTSans-Regular",
          "fontSize": 16,
          "lineHeight": 18,
        }
      }
    >
      John Doe
    </Text>
    <Text
      kind="black"
      style={
        {
          "color": "#FFFFFF",
          "fontFamily": "PTSans-Regular",
          "fontSize": 14,
          "lineHeight": 16,
          "textAlign": "center",
        }
      }
    >
      **** **** **** 8727
    </Text>
    <Text
      kind="black"
      style={
        {
          "color": "#FFFFFF",
          "fontFamily": "PTSans-Regular",
          "fontSize": 14,
          "lineHeight": 16,
          "textAlign": "center",
        }
      }
    >
      validity 12/23
    </Text>
  </View>
</View>
`;
